File: modules/angular2/src/web-workers/ui/event_serializer.ts
import {StringMap, Set} from 'angular2/src/facade/collection';

const MOUSE_EVENT_PROPERTIES = [
  "altKey",
  "button",
  "clientX",
  "clientY",
  "metaKey",
  "movementX",
  "movementY",
  "offsetX",
  "offsetY",
  "region",
  "screenX",
  "screenY",
  "shiftKey"
];

const KEYBOARD_EVENT_PROPERTIES = [
  'altkey',
  'charCode',
  'code',
  'ctrlKey',
  'isComposing',
  'key',
  'keyCode',
  'location',
  'metaKey',
  'repeat',
  'shiftKey',
  'which'
];

const EVENT_PROPERTIES = ['type', 'bubbles', 'cancelable'];

const NODES_WITH_VALUE =
    new Set(["input", "select", "option", "button", "li", "meter", "progress", "param"]);

export function serializeGenericEvent(e: Event): StringMap<string, any> {
  return serializeEvent(e, EVENT_PROPERTIES);
}

// TODO(jteplitz602): Allow users to specify the properties they need rather than always
// adding value #3374
export function serializeEventWithValue(e: Event): StringMap<string, any> {
  var serializedEvent = serializeEvent(e, EVENT_PROPERTIES);
  return addValue(e, serializedEvent);
}

export function serializeMouseEvent(e: MouseEvent): StringMap<string, any> {
  return serializeEvent(e, MOUSE_EVENT_PROPERTIES);
}

export function serializeKeyboardEvent(e: KeyboardEvent): StringMap<string, any> {
  var serializedEvent = serializeEvent(e, KEYBOARD_EVENT_PROPERTIES);
  return addValue(e, serializedEvent);
}

// TODO(jteplitz602): #3374. See above.
function addValue(e: Event, serializedEvent: StringMap<string, any>): StringMap<string, any> {
  if (NODES_WITH_VALUE.has((<HTMLElement>e.target).tagName.toLowerCase())) {
    serializedEvent['target'] = {'value': (<HTMLInputElement>e.target).value};
  }
  return serializedEvent;
}

function serializeEvent(e: any, properties: List<string>): StringMap<string, any> {
  var serialized = {};
  for (var i = 0; i < properties.length; i++) {
    var prop = properties[i];
    serialized[prop] = e[prop];
  }
  return serialized;
}

File: modules/angular2/src/web-workers/ui/event_serializer.dart
library angular2.src.web_workers.event_serializer;

import 'package:angular2/src/facade/collection.dart';
// TODO(jteplitz602): Remove Mirrors from serialization #3348
@MirrorsUsed(
    symbols: "altKey, bubbles, button, cancelable, client, ctrlKey, " +
        "defaultPrevented, detail, eventPhase, layer, metaKey, offset, page, region, screen, " +
        "shiftKey, timeStamp, type, magnitude, x, y, charCode, keyCode, keyLocation, location, repeat")
import 'dart:mirrors';
import 'dart:core';
import 'dart:html';

// These Maps can't be const due to a dartj2 bug (see http://github.com/dart-lang/sdk/issues/21825)
// Once that bug is fixed these should be const
final Map MOUSE_EVENT_PROPERTIES = {
  #altKey: bool,
  #bubbles: bool,
  #button: int,
  #cancelable: bool,
  #client: Point,
  #ctrlKey: bool,
  #defaultPrevented: bool,
  #detail: int,
  #eventPhase: int,
  #layer: Point,
  #metaKey: bool,
  #offset: Point,
  #page: Point,
  #region: String,
  #screen: Point,
  #shiftKey: bool,
  #timeStamp: int,
  #type: String
};

final Map KEYBOARD_EVENT_PROPERTIES = {
  #altKey: bool,
  #bubbles: bool,
  #cancelable: bool,
  #charCode: int,
  #ctrlKey: bool,
  #defaultPrevented: bool,
  #detail: int,
  #eventPhase: int,
  #keyCode: int,
  #keyLocation: int,
  #layer: Point,
  #location: int,
  #repeat: bool,
  #shiftKey: bool,
  #timeStamp: int,
  #type: String
};

Map<String, dynamic> serializeMouseEvent(dynamic e) {
  return serializeEvent(e, MOUSE_EVENT_PROPERTIES);
}

Map<String, dynamic> serializeKeyboardEvent(dynamic e) {
  return serializeEvent(e, KEYBOARD_EVENT_PROPERTIES);
}

Map<String, dynamic> serializeEvent(dynamic e, Map<Symbol, Type> PROPERTIES) {
  var serialized = StringMapWrapper.create();
  var mirror = reflect(e);
  PROPERTIES.forEach((property, type) {
    var value = mirror.getField(property).reflectee;
    var propertyName = MirrorSystem.getName(property);
    if (type == int || type == bool || type == String) {
      serialized[propertyName] = value;
    } else if (type == Point) {
      var point = reflect(value);
      serialized[propertyName] = {
        'magnitude': point.getField(#magnitude).reflectee,
        'x': point.getField(#x).reflectee,
        'y': point.getField(#y).reflectee
      };
    }
  });

  return serialized;
}
